### PR TITLE
Add extra options for handling pinned tabs in :tab-only

### PR DESCRIPTION
**Pull request by nichobi**
~~Adds a `--soft` flag to `:tab-only`, which keeps pinned tabs without prompting the user. Essentially it's the opposite of `--force`.~~

Replaces the `--force` flag for `:tab-only` with a `--pinned` flag which decides how pinned tabs should be handled, allowing for the following values: `prompt` (default), `close` and `keep`.

This is my first pull request for qutebrowser, so I'd appreciate any feedback or guidance. So far this is a barebones implementation without any tests.

My use case is that I pin tabs precisely so they won't get closed by `:tab-only`, so it becomes somewhat annoying to have to decline the prompt each time.

### CODE REVIEW

**Comment by The-Compiler** [comment]
That seems useful - I don't use pinned tabs myself at all, so I rely on users to tell me what's missing there :slightly_smiling_face: 

Let's first look at how the "API" of that command should look like to the user:

I guess with the `--soft` name, you got inspired by `--force`? I'd rather have a more descriptive name (say, `--keep-pinned` maybe?), even if `--force` also isn't very descriptive in this case.

Two more possibilities I can think of:

- Replacing `--force` entirely, with something like a `--pinned` flag, taking an argument like `close`, `ask` or `keep`.
- Using a setting instead of a flag instead - however, that's a bit less flexible, I think it makes more sense to keep it as a flag.

What do you think?

**Comment by nichobi** [comment]
Yeah, I was a bit unsure what to call it, `keep-pinned` would be certainly be an improvement if keeping it as a flag. 
I lean towards your `--pinned`suggestion, since it seems more flexible, and would also clear up the combination of `--force --soft` which currently would just ignore the `--soft` flag.
If I was to implement this, should I attempt (temporary) compatibility with the `--force` flag, or is it okay to just replace it entirely, breaking people's bindings?

**Comment by The-Compiler** [comment]
You can use something like `cmdutils.check_exclusive((force, soft), 'fs')` to make them mutually exclusive, similarly to what's already done for `--prev` and `--next` at the beginning of the function.

I'd still go for `--pinned` though - what we usually do for changes like this is to keep the old API around for a release, but with a deprecation warning. Something like (untested):

```python
if force:
    message.warning("--force is deprecated, use --pinned close instead.")
    pinned = 'close'
```

as well as a `@cmdutils.argument('force', hide=True)` decorator to hide it from the docs.

**Comment by nichobi** [comment]
```python
    @cmdutils.register(instance='command-dispatcher', scope='window')
    @cmdutils.argument('pinned', choices=['prompt', 'close', 'keep'])
    @cmdutils.argument('force', hide=True)
    def tab_only(self, prev=False, next_=False, pinned='prompt', force=False):
```
I've changed the definition to the above, but `pinned` seems to be turned into a positional argument, i.e. `:tab-only close`. Is it not possible have it as a flag when it takes a string argument, i.e. `:tab-only --pinned close`?

**Comment by The-Compiler** [comment]
That's because it's a positional argument in Python, but doesn't use a `False` default value (making it a flag).

You should be able to make the arguments explicitly [keyword only](https://python-3-for-scientists.readthedocs.io/en/latest/python3_advanced.html#keyword-only-arguments) by doing `self, *, prev=False, ...` and qutebrowser should do the right thing.

File: qutebrowser/browser/commands.py [hunk 1]
```diff
@@ -756,13 +756,18 @@ def _yank_to_target(self, s, sel, what, keep, quiet):
...
+    def tab_only(self, *, prev=False, next_=False, pinned='prompt', force=False):
         """Close all tabs except for the current one.
 
         Args:
             prev: Keep tabs before the current.
             next_: Keep tabs after the current.
+            pinned: What to do with pinned tabs.
+                    Valid values: prompt, close, keep.
             force: Avoid confirmation for pinned tabs.
+            soft: Keep pinned tabs without prompting.
```
Comment (The-Compiler) [inline qutebrowser/browser/commands.py:770]: This can be removed now.

**Review by The-Compiler — CHANGES_REQUESTED** [review]
Thanks for the update! I think the only missing thing now is a test - search for `Pinned :tab-only` in `tests/end2end/features/tabs.feature` for some inspiration.

File: qutebrowser/browser/commands.py (hunk 1 above)
Comment (nichobi) [reply qutebrowser/browser/commands.py]: Oops! I've amended that commit

File: qutebrowser/browser/commands.py [hunk 2]
```diff
@@ -775,19 +779,23 @@ def _to_close(i):
...
+                if pinned == 'close' or not tab.data.pinned:
                     self._tabbed_browser.close_tab(tab, new_undo=first_tab)
                     first_tab = False
                 else:
                     pinned_tabs_cleanup = tab
 
         # Check to see if we would like to close any pinned tabs
-        if pinned_tabs_cleanup:
+        if pinned_tabs_cleanup and pinned == 'prompt':
             self._tabbed_browser.tab_close_prompt_if_pinned(
```
Comment (The-Compiler) [inline qutebrowser/browser/commands.py:799]: In the lambda below this line (GitHub won't let me comment there as it's not part of your changes), `force=True` is still used, thus triggering the warning (and failing `test_pinned_tabonly_prompt_yes`). If you change that to `pinned='close'` instead, I think everything should work fine.

**Review by The-Compiler — CHANGES_REQUESTED** [review]

**Comment by The-Compiler** [comment]
Looks great now, thanks! Pushed with some minimal changes to the docs in 103017876e3ad3a74591fdd5053addb11a43f8f5.

**Comment by nichobi** [comment]
Thank you so much for all the help! It's made the codebase feel much less intimidating, so hopefully my next contribution will go smoother!

**Pull request merged**